lib/sshkit/ssh/connection.ex

```diff
@@ -42,27 +46,51 @@ defmodule SSHKit.SSH.Connection do
     open(to_charlist(host), options)
   end
   def open(host, options) do
-    port = Keyword.get(options, :port, 22)
-    timeout = Keyword.get(options, :timeout, :infinity)
-    ssh_modules = Keyword.get(options, :ssh_modules, @ssh_modules)
-    ssh = erlang_module(%{ssh_modules: ssh_modules}, :ssh)
+    {ssh_options, sshkit_options} = fetch_options(options)
+
+    ssh = ssh_module(sshkit_options)
+    case ssh.connect(host, sshkit_options.port, ssh_options, sshkit_options.timeout) do
+      {:ok, ref} -> {
+        :ok,
+        %Connection{
+          host: host,
+          port: sshkit_options.port,
+          options: ssh_options,
+          ref: ref,
+          ssh_modules: sshkit_options.ssh_modules
+        }
+      }
+      err -> err
+    end
+  end
 
-    defaults = [user_interaction: false]
+  @default_erlang_ssh_options [user_interaction: false]
+  defp fetch_options(options) do
+    dry_run = Keyword.get(options, :dry_run, false)
+    ssh_modules = if dry_run do
+      @dry_run_ssh_modules
+    else
+      @ssh_modules
+    end
 
-    options =
-      defaults
+    sshkit_options = %{
+      port: Keyword.get(options, :port, 22),
+      timeout: Keyword.get(options, :timeout, :infinity),
+      ssh_modules: Keyword.get(options, :ssh_modules, ssh_modules),
```

Comment:
right! and this was intended. Because the `ssh_modules` option is an undocumented feature I don't expect users to set it (and if they do they do it intentionally, like we do in our tests with the mock-modules).

we should start to worry when we have multiple features fighting for their `ssh_module` to be set-up. But that's probably a problem for our future selfs :)

![that's a job for future-me](http://cdn.pidjin.net/wp-content/uploads/2013/10/cover-fb.png)